semgrep-core/src/parsing/pfff/ruby_to_generic.ml

```diff
@@ -177,6 +177,11 @@ and argument arg : G.argument =
   | Arg (Binop (Atom (_tk, AtomSimple id), (Op_ASSOC, _v2), v3)) ->
       let e3 = expr v3 in
       G.ArgKwd (id, e3)
+  (* Ruby allows surrounding single arguments with useless parenthesis *)
```

Comment:
I did a little investigating, and it's more complicated than this. In Ruby you can omit parentheses in a function call, so that `f(x)` and `f x` are equivalent. The problem now is that when someone writes `f (x)`, they want whitespace to be irrelevant: they mean `f(x)`. But if `f x` = `f(x)`, then `f (x) = f((x))`, and `f((x)) != f(x)`. This is because `f(x if false) != f((x if false))` (see https://stackoverflow.com/questions/17036166/why-do-i-need-extra-parentheses-for-conditional-modifier for why).

I'm not entirely sure what I think is the right solution here tbh; I _think_ using `ParenExpr` (to preserve parentheses for autofix) is still best, but it is a little weird since by the time this gets to Semgrep we don't care if it's `f x`, `f(x)`, or `f((x))`. Another option is to specifically change tree-sitter-ruby to parse `f (x)` as equivalent to `f(x)` (I don't _think_ there are any cases of valid code where that would matter). I don't prefer it mostly because I'm not comfortable enough with Ruby to say whether there would be unintended effects.

Reply:
At the very least, the requested changes is to explain all this in the comment :D